Add group#new & group#create function

tokuchan-pg opened the pull request:
# WHAT
グループ新規作成機能の実装

# WHY
グループ機能実装のため

### グループ新規作成画面↓
![2017-04-22 20 40 20](https://cloud.githubusercontent.com/assets/25572309/25304137/1d2ab92a-279c-11e7-8370-8fdac5b34d9c.png)

tokuchan-pg committed:
generate joins table  with foreign key constrains

Joins table is a middle table of users table and groups table.
Type of foreign key is references in rails.

tokuchan-pg committed:
defined association betweet group model and user model

tokuchan-pg committed:
defined accepts_nested_attributes_for in group model

To save records in groups table and joins table(middle table of groups
and users) at the same time.

tokuchan-pg committed:
defined new & create action in groups_controller

tokuchan-pg committed:
replaced form tag to form_for helper method

tokuchan-pg committed:
replaced input tag to collection_check_boxes method

ninooo29 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,6 +1,18 @@
+    Group.create(group_params)
+    redirect_to root_path
+  end
+
```
失敗時の処理を書きましょう。

tokuchan-pg replied:
createアクションを編集し、失敗時の以下の実装を追加しました
- フラッシュメッセージを表示させる
- renderを使って、group#newのビューを表示させる（createアクションの中でも@groupを定義し、renderを使うことで、Saveボタンを押す前の入力内容が保持される）

ninooo29 commented on app/views/groups/new.html.haml:
```diff
@@ -23,7 +27,8 @@
           #chat-group-user-22.chat-group-user
             %input{:name => "chat_group[user_ids][]", :type => "hidden", :value => "22"}/
             %p.chat-group-user__name seo_kyohei
```
ログインしているユーザーの名前が出ているようにしましょう。

ninooo29 commented on app/views/groups/new.html.haml:
```diff
@@ -24,9 +24,8 @@
-            %input{:name => "chat_group[user_ids][]", :type => "hidden", :value => "22"}/
-            %p.chat-group-user__name seo_kyohei
+          .chat-group-user
+            = collection_check_boxes(:group, :user_ids, User.all, :id, :name)
```
ログイン中のユーザーの名前はチェックボックスに表示されないようにしましょう。

tokuchan-pg replied:
ログインユーザーのチェックボックスを非表示にしました。同時に、groupコントローラーを編集して、ログインユーザーを中間テーブルに自動的に保存されるようにしました。

ninooo29 commented on app/models/group.rb:
```diff
@@ -1,2 +1,6 @@
 class Group < ApplicationRecord
+
+  has_many :joins
+  has_many :users, through: :joins
```
中間テーブルの命名は命名規則よりgroup_usersに変更おねがいいたします。

tokuchan-pg replied:
rollbackにより、中間テーブルの名前を変更いたしました。これに伴い、以下の変更も加えました。（DB設計自体に変更はありません。）
- アソシエーションの記述を変更
- モデルのファイル名を変更
- モデルのクラス名を変更

ninooo29 commented on app/models/user.rb:
```diff
@@ -4,6 +4,9 @@ class User < ApplicationRecord
          :recoverable, :rememberable, :trackable, :validatable
 
+  has_many :joins
+  has_many :groups, through: :joins
```
ここも変更お願いします！

tokuchan-pg replied:
対応致しました。

tokuchan-pg committed:
renamed the middle table from 'joins' to 'group_users'

Used ‘rake db:rollback’ to rename. There are some related changes at
the same time.

*changed the name of the migration file and the class name of it

*changed the table name in association setting in model files

*changed the name of the middle table’s model file and the class name
of it

*revised the README.md

There is no change in table structure. Just changed the name of it.

tokuchan-pg committed:
hide checkbox of login-user from view

At the same time, add the login-user automatically to the new group

tokuchan-pg committed:
added failure pattern to group#create action

Condition is defined based on the result of validation in
models/group.rb.
Also, show flash messages in both success and failure pattern.

tokuchan-pg commented:
## 以下の修正を行いました

**1. 中間テーブルの名前の変更**
これに伴い、以下の変更も加えています。DB設計自体に変更はありません。  
    - アソシエーションの記述の変更
    - モデルのファイル名とクラス名の変更

**2. group#newのビューにて、メンバー選択画面からログインユーザーのチェックボックスを非表示に**
同時に、ログインユーザーのidがgroup#createアクションに自動的に渡されるように設定しました。
  
**3. group#createアクションに成功時と失敗時の分岐を追加**
models/group.rbに設定したvalidation（グループの新規作成時にnameの入力を必須にする）の結果に基づいて、成功パターンと失敗パターンに分岐させました。
    - 成功時はトップページにリダイレクト
    - 失敗時はrenderを使って、Saveボタンが押される前の入力内容を保持した状態でgroup#newのビューを表示
    - 両パターンにそれぞれ異なるフラッシュメッセージを追加
  
#### group#createアクション成功時のフラッシュメッセージ↓
![2017-04-23 16 56 19](https://cloud.githubusercontent.com/assets/25572309/25311982/7293b9be-2848-11e7-92ac-17816eeb23fb.png)
  
#### group#createアクション失敗時のフラッシュメッセージ↓
![2017-04-23 16 56 07](https://cloud.githubusercontent.com/assets/25572309/25311981/6f4a3c74-2848-11e7-9939-91776d26c884.png)

sato11 commented on app/views/groups/new.html.haml:
```diff
@@ -1,29 +1,34 @@
-  %form#new_chat_group.new_chat_group{"accept-charset" => "UTF-8", :action => "/chat_groups", :method => "post"}
-    %input{:name => "utf8", :type => "hidden", :value => "✓"}/
-    %input{:name => "authenticity_token", :type => "hidden", :value => "AxFKlYEhD6eqX1PiZoTYQDANtKvgcFSXZQXHt5hyTl55/U/CHC5rtPavrcu2za65riz4tekZEy56vP6kEb2wYA=="}/
+  =form_for @group, html: {id: 'new_chat_group', class: 'new_chat_group'} do |f|
```
ヘルパーメソッド `form_for` はあえて指定せずとも最適化されたクラスと id を付与してくれるので、オプションの指定は必要ありません。

tokuchan-pg replied:
修正いたしました。修正前にhtmlオプションで指定していたセレクタは、group.scssの中で使われていないことも確認いたしました。

sato11 commented on app/views/groups/new.html.haml:
```diff
@@ -1,29 +1,34 @@
-    %input{:name => "authenticity_token", :type => "hidden", :value => "AxFKlYEhD6eqX1PiZoTYQDANtKvgcFSXZQXHt5hyTl55/U/CHC5rtPavrcu2za65riz4tekZEy56vP6kEb2wYA=="}/
+  =form_for @group, html: {id: 'new_chat_group', class: 'new_chat_group'} do |f|
+
+    / グループ名入力
```
こちらはメモでしょうか？
わかりにくいコメントほど後になって困るものはないので、削除するか詳細に記述するかのどちらかを選びましょう。
個人的にはこの程度ならコードを読めばわかるはずなので、必要ないかなと思います。

tokuchan-pg replied:
修正いたしました。第三者にも配慮した記述を心がけます。

sato11 commented on app/views/groups/new.html.haml:
```diff
@@ -1,29 +1,34 @@
-        .chat-group-form__search
-          %input#user-search-field.chat-group-form__input{:placeholder => "追加したいユーザー名を入力してください", :type => "text"}/
-        #user-search-result
+        =f.text_field :name, placeholder: 'グループ名を入力してください', id: 'chat_group_name', class: 'chat-group-form__input'
```
こちらもセレクタの記述は必要ないのではないでしょうか？

tokuchan-pg replied:
修正いたしました。修正前に自分で指定していたセレクタがgroup.scssの中で使われていたので、group.scssのセレクタの記述を自動生成されたものに修正いたしました。

sato11 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,6 +1,26 @@
+      flash.notice = '新規グループが作成されました。'
+      redirect_to root_path
+    else
+      flash.alert = 'グループ名を入力して下さい'
```
失敗後にフラッシュメッセージが出るのは確認いただいたかと思いますが、その後続けて画面を回遊すると必要以上に渡ってフラッシュが画面上に残ってしまいませんか？

tokuchan-pg replied:
flash.nowを使って修正いたしました。
flashは「アクション->リダイレクト->アクション」の処理が済まないとメッセージが削除されない仕様なので、アクションが動かないrenderと併用すると、ページを回遊してもメッセージが残ってしまうのが原因だったと推察されます。

sato11 commented on app/controllers/groups_controller.rb:
```diff
@@ -1,6 +1,26 @@
+    # ログインユーザーのidを、collection_check_boxes経由で送られてきた配列user_idsに、文字列型で追加する
+    params.require(:group)[:user_ids].push(current_user.id.to_s)
+    # :user_idsは配列なので、書き方が↓のように特殊な形となる
+    return params.require(:group).permit(:name, :user_ids => [])
```
やや冗長気味なので綺麗に書きたいですね。

```
def group_params
  params = params.require(:group).permit(:name, user_ids: [])
  params[:user_ids].push(current_user.id.to_s)
  params
end
```

とか

```
def group_params
  params[:group][:user_ids].push(current_user.id.to_s)
  params.require(:group).permit(:name, user_ids: [])
end
```

といったようにならないでしょうか？
上記、実際には実行していないので鵜呑みにせずに確認してみてくださると嬉しいです。

tokuchan-pg replied:
修正いたしました。
ご紹介いただいた２つ目のコードが、自分のコードの文脈と近しかったので、参考にさせていただきました。挙動も確認いたしました。

sato11 commented on app/models/group.rb:
```diff
@@ -1,2 +1,12 @@
 class Group < ApplicationRecord
+
+  # グループ名が入力されていない時はテーブルにレコードを保存しない
+  validates :name, presence: true
```
内容とは関係ないのですが、コードを読めば自明であるので、このようなコメントは必ずしも必要ではないかなと思います。
コメントは多ければ多いほど親切なようでいて、実は多すぎるコメントは却って読解のスピードを遅めてしまわないかなと思います。
もちろん複雑に入り組んだ箇所などには積極的にコメントしていくべきとは思いますが、この程度であれば不要というのが僕の考えです。いかがでしょうか？

tokuchan-pg replied:
修正いたしました。

確かにコメントが多すぎるのも煩雑になってしまうと反省しました。レビューしてくださる方が煩わしさを感じない記述をもっと心がけたいと思います！

tokuchan-pg committed:
deleted id & class option from from_for helper

Because, form_for helper generate id & class name automatically.

tokuchan-pg committed:
deleted id & class name option from f.text_field

Because ‘f’ generate id & class name automatically.
At the same time, revised the selector in groups.scss

tokuchan-pg committed:
show flash message only once when group#create failed

When ‘flash’ method is used with ‘render’ method,  flash message
appears more than once. So replaced into ‘flash.now’.

tokuchan-pg committed:
made the code of the strong parameter more simple

There is no change in its function.

tokuchan-pg committed:
deleted unnecessary comment

vip-take commented on app/controllers/groups_controller.rb:
```diff
@@ -10,7 +10,10 @@ def create
+      # renderと併用すると、ページを回遊してもメッセージが残ってしまう。
+      # →flash.nowを使う
+      flash.now.alert = 'グループ名を入力して下さい'
       render action: 'new'
```
このように簡潔に書くことも可能ですので、直しましょう！

```
render :new
```

tokuchan-pg replied:
修正いたしました。

vip-take commented on app/controllers/groups_controller.rb:
```diff
@@ -10,7 +10,10 @@ def create
-      flash.alert = 'グループ名を入力して下さい'
+      # flashは「アクション->リダイレクト->アクション」の処理が済まないとメッセージが削除されないので、
+      # renderと併用すると、ページを回遊してもメッセージが残ってしまう。
+      # →flash.nowを使う
```
とても親切なコメントですが、それなりに動作を理解している人にとってはあまり有益でないコメントかと思いますので、コードへの記載は控えましょう！
（私も自分用メモみたいな感じで書きたくなる時あるんですけどね。。笑）

tokuchan-pg replied:
修正いたしました。

vip-take commented on app/controllers/groups_controller.rb:
```diff
@@ -21,9 +21,9 @@ def create
     # :user_idsは配列なので、書き方が↓のように特殊な形となる
-    return params.require(:group).permit(:name, :user_ids => [])
+    params.require(:group).permit(:name, user_ids: [])
   end
```
[参考]
createアクションでnewする際に以下のように書いてあげることもできます。
```
current_user.groups.new(group_params)
```

tokuchan-pg committed:
code omission

tokuchan-pg committed:
delete unnecessary comment
